protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java

```diff
@@ -0,0 +1,305 @@
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.remove(messageUid1);
+
+        assertThat(testee.getMsn(messageUid1).isPresent()).isFalse();
+        assertThat(testee.getMsn(messageUid2).isPresent()).isTrue();
+        assertThat(testee.getMsn(messageUid3).isPresent()).isTrue();
+    }
+
+    @Test
+    public void  removeShouldRemoveOnlyTheCorrectUidWhenOneTheEnd() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.remove(messageUid3);
+
+        assertThat(testee.getMsn(messageUid3).isPresent()).isFalse();
+        assertThat(testee.getMsn(messageUid1).isPresent()).isTrue();
+        assertThat(testee.getMsn(messageUid2).isPresent()).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheFirstOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid1);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheLastOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid4);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingOneInTheMiddle() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid3);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingTheFirstOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid1);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingTheLastOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid4);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingOneInTheMiddle() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid3);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    private static boolean messageNumberAreContiguous(UidMsnMapper testee) {
+        int first = UidMsnMapper.FIRST_MSN;
```

Comment:
1: Use ate least Guava for this.
2: You have 4 UIDs. Please replace these method by more asserts. You ca IE make asserts on collections with a method @VisibleForTesting